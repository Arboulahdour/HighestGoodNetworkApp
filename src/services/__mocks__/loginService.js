let getCurrentUserResult;
const getCurrentUseroptions = {
  userPresent: { foo: "bar", baz: "masklsd" },
  userNotPresent: null,
  headerTest: {
    name: "Foobar",
    profilePic: undefined,
    userId: "5be0952c633dae0016081b4b"
  }
};

export function getCurrentUser() {
  return getCurrentUserResult;
}

export function login() {
  return "Success";
}

getCurrentUser.__setValue = option => {
  getCurrentUserResult = getCurrentUseroptions[option];
};
