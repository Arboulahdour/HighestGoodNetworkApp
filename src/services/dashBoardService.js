import httpService from './httpService';

// const ApiEndpoint = `https://hgn-rest-dev.herokuapp.com/api/dashboard`;
const ApiEndpoint = `${process.env.REACT_APP_APIENDPOINT}/dashboard`;


export function getLeaderboardData(userId) {
  return httpService.get(`${ApiEndpoint}/leaderboard/${userId}`);
}
