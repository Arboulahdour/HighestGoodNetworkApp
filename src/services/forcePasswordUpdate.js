// import httpService from "./httpService";

// const ApiEndpoint = `${process.env.REACT_APP_APIENDPOINT}/forcepassword`;

// export function forcePasswordUpdate(data) {
//   return httpService.patch(ApiEndpoint, data);
// }
