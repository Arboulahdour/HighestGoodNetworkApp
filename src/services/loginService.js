import jwtDecode from 'jwt-decode';
import httpService from './httpService';
import config from '../config.json';

const loginApiEndpoint = `${process.env.REACT_APP_APIENDPOINT}/login`;

const { tokenKey } = config;

// export async function login(credentials) {
//   const { data } = await httpService.post(loginApiEndpoint, credentials);
//   if (data.new) {
//     return { userType: 'newUser', userId: data.userId };
//   }
//   localStorage.setItem(tokenKey, data.token);
// }

// export function logout() {
//   localStorage.removeItem(tokenKey);
// }

// export function getCurrentUser() {
//   return isUserAuthenticated()
//     ? jwtDecode(localStorage.getItem(tokenKey))
//     : null;
// }
// export function loginWithJWT(jwt) {
//   localStorage.setItem(tokenKey, jwt);
// }

// export function getjwt() {
//   return localStorage.getItem(tokenKey) ? localStorage.getItem(tokenKey) : null;
// }

// export function isUserAuthenticated() {
//   if (!localStorage.getItem(tokenKey)) {
//     return false;
//   }
//   const token = localStorage.getItem(tokenKey);
//   const { expiryTimestamp } = jwtDecode(token);
//   return expiryTimestamp > new Date().toISOString();
// }
