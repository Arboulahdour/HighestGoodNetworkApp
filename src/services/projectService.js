import httpService from './httpService';

const ApiEndpoint = `${process.env.REACT_APP_APIENDPOINT}/projects`;


export function getAllProjects() {
  return httpService.get(`${ApiEndpoint}`);
}
