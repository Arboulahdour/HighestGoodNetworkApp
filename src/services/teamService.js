import httpService from "./httpService";

const ApiEndpoint = `${process.env.REACT_APP_APIENDPOINT}/team`;

export function getAllTeams() {
  return httpService.get(`${ApiEndpoint}`);
}
