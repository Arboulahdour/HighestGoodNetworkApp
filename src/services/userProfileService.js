import httpService from './httpService'

const APIEndPoint = `${process.env.REACT_APP_APIENDPOINT}/userprofile`

// export function getUserProfile(userId) {
// 	return httpService.get(`${APIEndPoint}/${userId}`)
// }

// export function editUserProfileData(user, userId) {}

// export function postUserProfileData(user) {}

// export function updatePassword(userId, newpassworddata) {
// 	return httpService.patch(`${APIEndPoint}/${userId}/updatePassword`, newpassworddata)
// }
