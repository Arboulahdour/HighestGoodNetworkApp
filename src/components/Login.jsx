import React, { Component } from "react";
import Joi from "joi";
import { renderInput, renderButton } from "./common/form";
import { login, getCurrentUser } from "../services/loginService";
import { Redirect } from "react-router-dom";

class Login extends Component {
  state = {
    data: { email: "", password: "" },
    errors: {}
  };

  schema = {
    email: Joi.string()
      .email()
      .required()
      .label("Email"),
    password: Joi.string()
      .required()
      .label("Password")
  };

  doSubmit = async () => {
    const email = this.state.data.email;
    const password = this.state.data.password;
    try {
      let result = await login({ email, password });
      if (result && result.userType === "newUser") {
        window.location = `/forcePasswordUpdate/${result.userId}`;
        return;
      }
      const { state } = this.props.location;
      window.location = state ? state.from.pathname : "/dashboard";
      return;
    } catch (ex) {
      if (ex.response && ex.response.status === 403) {
        const errors = this.state.errors;

        errors["email"] = ex.response.data.message;
        this.setState({ errors });
      }
    }
  };

  render() {
    if (getCurrentUser()) return <Redirect to="/" />;

    return (
      <div className="container mt-5">
        <h2>Please Sign in</h2>

        <form className="col-md-6 xs-12" onSubmit={e => this.handleSubmit(e)}>
          {renderInput("email", "Email:")}
          {renderInput("password", "Password:", "password")}
          {renderButton("Submit")}
        </form>
      </div>
    );
  }
}

export default Login;
