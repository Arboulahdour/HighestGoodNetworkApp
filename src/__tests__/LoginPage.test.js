import React from "react";
import { shallow, mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import Login from "../components/Login";

import { getCurrentUser } from "../services/loginService";

jest.mock("../services/loginService");

describe("Login page structure", () => {
  let mountedLogin;
  beforeEach(() => {
    getCurrentUser.__setValue("userNotPresent");
    mountedLogin = shallow(<Login />);
  });

  it("should be rendered with two input fields", () => {
    const inputs = mountedLogin.find("Input");
    expect(inputs.length).toBe(2);
  });

  it("should be rendered with one button", () => {
    const button = mountedLogin.find("button");
    expect(button.length).toBe(1);
  });
  it("should be rendered with one h2", () => {
    const h2 = mountedLogin.find("h2");
    expect(h2.length).toEqual(1);
    expect(h2.first().text()).toContain("Please Sign in");
  });
  it("should match the snapshot", () => {
    expect(mountedLogin).toMatchSnapshot();
  });
});

describe("When user tries to input data", () => {
  let mountedLoginPage;

  beforeEach(() => {
    getCurrentUser.__setValue("userNotPresent");
    mountedLoginPage = shallow(<Login />);
  });

  it("should call handleInput when input is changed", () => {
    const spy = jest.spyOn(mountedLoginPage.instance(), "handleInput");
    mountedLoginPage
      .find("[name='email']")
      .simulate("change", { currentTarget: { name: "email", value: "abc" } });
    expect(spy).toHaveBeenCalled();
  });

  it("should correctly update the email value in the state", () => {
    const expected = "sh@gmail.com";
    const Input = { name: "email", value: expected };
    const mockEvent = { currentTarget: Input };
    mountedLoginPage.instance().handleInput(mockEvent);

    expect(mountedLoginPage.instance().state.data.email).toEqual(expected);
  });

  it("should correctly update the error if the email is not in correct format", () => {
    const expected = "sh";
    const Input = { name: "email", value: expected };
    const mockEvent = { currentTarget: Input };
    mountedLoginPage.instance().handleInput(mockEvent);
    expect(mountedLoginPage.instance().state.errors.email).toEqual(
      '"Email" must be a valid email'
    );
  });

  it("should correctly update the password value in the state", () => {
    const expected = "trapp";
    const Input = { name: "password", value: expected };
    const mockEvent = { currentTarget: Input };
    mountedLoginPage.instance().handleInput(mockEvent);

    expect(mountedLoginPage.instance().state.data.password).toEqual(expected);
  });

  it("should correctly update the errors object if the password is empty", () => {
    const expected = "";
    const Input = { name: "password", value: expected };
    const mockEvent = { currentTarget: Input };
    mountedLoginPage.instance().handleInput(mockEvent);
    expect(mountedLoginPage.instance().state.errors.password).toEqual(
      '"Password" is not allowed to be empty'
    );
  });

  it("should have disabled submit button if form is invalid", () => {
    const button = mountedLoginPage.find("button");
    expect(button.props()).toHaveProperty("disabled");
  });

  it("form can be submitted", () => {
    const callback = jest.fn();
    const mountedLoginPagewithCallBack = shallow(<form onSubmit={callback} />);
    mountedLoginPagewithCallBack.find("form").simulate("submit");
    expect(callback).toHaveBeenCalled();
  });

  it("onSubmit login method is called", async () => {
    const loginService = require("../services/loginService");
    const syplogin = jest.spyOn(loginService, "login");
    await mountedLoginPage.instance().doSubmit();
    expect(syplogin).toHaveBeenCalled();
  });
});

describe("Login behavior", () => {
  it("should have redirection set to homepage ", () => {
    getCurrentUser.__setValue("userPresent");
    const wrapper = mount(
      <MemoryRouter>
        <Login />
      </MemoryRouter>
    );

    const redirect = wrapper.find("Redirect");
    expect(redirect.props()).toHaveProperty("to", "/");
  });

  xit("should perform correct redirection if user was redirected to login page from some other location", async () => {
    getCurrentUser.__setValue("userPresent");

    const somepath = "/profile/1234";
    const location = { state: { from: { pathname: somepath } } };
    const wrapper = shallow(
      <MemoryRouter>
        <Login location={location} />
      </MemoryRouter>
    );
    expect(wrapper.instance().props.location.state.from.pathname).toEqual(
      somepath
    );
    wrapper.instance().doSubmit();

    expect(window.location.pathname).toEqual(somepath);
  });

  it("should populate errors if login fails", async () => {
    const errorMessage = { message: "Invalid email and/ or password." };

    const loginService = require("../services/loginService");
    loginService.login = jest.fn(() => {
      throw { response: { status: 403, data: errorMessage } };
    });

    const mountedLoginPage = shallow(<Login />);
    await mountedLoginPage.instance().doSubmit();
    expect(mountedLoginPage.instance().state.errors.email).toEqual(
      errorMessage.message
    );
  });
});
