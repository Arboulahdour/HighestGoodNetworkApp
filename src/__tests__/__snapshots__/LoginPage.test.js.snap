// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page structure should match the snapshot 1`] = `undefined`;
